### PR TITLE
C++ Vocabulary

### PR DESCRIPTION
**Pull request by ShahriarSS**
Hi @zhangguanheng66 Here is the vocabulary.

### CODE REVIEW

File: torchtext/csrc/core/vocabulary.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+#include <stdint.h>
+#include <string>
+#include <vector>
+
+namespace torchtext {
+namespace core {
+class Vocabulary {
+  static const size_t MAX_VOCAB_SIZE = 30000000;
+  static const size_t negative = static_cast<size_t>(-1);
```
Comment (cpuhrsch) [inline torchtext/csrc/core/vocabulary.h]: Why is this necessary?

File: torchtext/csrc/core/vocabulary.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
+#ifndef VOCABULARY_H
+#define VOCABULARY_H
+
+#include <stdint.h>
```
Comment (cpuhrsch) [inline torchtext/csrc/core/vocabulary.h]: Why not use the cpp compatibility header https://en.cppreference.com/w/cpp/header/cstdint

File: torchtext/csrc/core/vocabulary.cpp [hunk 3]
```diff
@@ -0,0 +1,37 @@
+#include "vocabulary.h"
+#include <iostream>
+
+namespace torchtext {
+namespace core {
+size_t Vocabulary::word_hash(const std::string& word) {
+  return std::hash<std::string>()(word);
```
Comment (cpuhrsch) [inline torchtext/csrc/core/vocabulary.cpp]: Do you know what hash function this uses? I'd prefer https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function

File: setup.py [hunk 4]
```diff
@@ -25,6 +32,55 @@ def find_version(*file_paths):
 VERSION = find_version('torchtext', '__init__.py')
 long_description = read('README.rst')
 
+
+def get_extensions():
```
Comment (cpuhrsch) [inline setup.py]: I assume this is copy paste from torchvision?

File: .clang-format [hunk 5]
```diff
@@ -0,0 +1,88 @@
+---
```
Comment (cpuhrsch) [inline .clang-format]: I'd want to add this in a separate PR.

**Comment by cpuhrsch** [comment]
Could you rename this from Vocabulary to Dictionary please :)

File: torchtext/csrc/core/vocabulary.h (hunk 1 above)
Comment (ShahriarSS) [reply torchtext/csrc/core/vocabulary.h]: Do you mean negative? I'm using ```size_t``` for indexes which is unsigned and indexes must initially be negative(-1). So this is represents negative in unsigned integer (std::string uses the same method for negative).

File: torchtext/csrc/core/vocabulary.cpp (hunk 3 above)
Comment (ShahriarSS) [reply torchtext/csrc/core/vocabulary.cpp]: No I don't but it can't be a bad one. Can it? I'll look into Fowler hash function.

File: setup.py (hunk 4 above)
Comment (ShahriarSS) [reply setup.py]: Yes. I'm not an expert in Python so I copied and changed to fit.

File: .clang-format (hunk 5 above)
Comment (ShahriarSS) [reply .clang-format]: OK.

File: torchtext/csrc/core/vocabulary.cpp (hunk 3 above)
Comment (ShahriarSS) [reply torchtext/csrc/core/vocabulary.cpp]: @cpuhrsch I looked into it and apparently GCC is using MurmurHashUnaligned2 (see [this](https://stackoverflow.com/questions/19411742/what-is-the-default-hash-function-used-in-c-stdunordered-map)). Maybe it would be better if we implemented the hash function ourselves to be sure that it will perform the same everywhere.

Comment (cpuhrsch) [reply torchtext/csrc/core/vocabulary.cpp]: Yes ideally we implement it ourselves, because this hash function also has particular properties for text.

Comment (ShahriarSS) [reply torchtext/csrc/core/vocabulary.cpp]: I did implement it.

File: torchtext/csrc/core/dictionary.cpp [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  auto words_size = words.size();
+  auto index = word_hash(word) % words_size;
+
+  while (words[index].index != negative && words[index].word != word)
+    index = (index + 1) & words_size;
+
+  return index;
+}
+
+size_t Dictionary::add_word(const std::string& word) {
```
Comment (cpuhrsch) [inline torchtext/csrc/core/dictionary.cpp]: After adding words, at some point, the user will ideally want to sort the dataset based on the frequency. The most frequent word should be the index 0, the second most frequent index 1, etc. - This is a memory optimization. Words follow [Zipf's law](https://en.wikipedia.org/wiki/Zipf%27s_law). When you read in a cacheline of data from this lookup you're most likely going to read index 0. Since you always read 64 bytes, that'll result in certain number of the first few words to always be read. The probability of hitting an index close to 0 is very high due to Zipf's law, so on the second read you'll likely be able to use the data you just read into cache. If the indicies on the other hand are random you'll constantly trash cache for a single word and read much more than you need to.

On that note, using std::string is probably not a good idea, because you need tight control over the structure that's used to represent the word. We can however add these optimizations later on once you have a benchmark (please add that as well) and can look at the memory pressure of a typical read. A good benchmark is tokenizing and creating a dictionary based on Wikipedia. At some point we might also want to align an entry struct or use SmallVector to store the first few characters on the stack.

**Comment by erip** [comment]
I think this is safe to close since we have a pybind and torchbind version of Vocab, cc @parmeet
